larpandoracontent/LArTwoDReco/LArCosmicRay/ClusterAssociation.h

```diff
@@ -0,0 +1,349 @@
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+inline bool ClusterAssociation::operator<(const ClusterAssociation &clusterAssociation) const
+{
+    return (LArClusterHelper::SortCoordinatesByPosition(m_upstreamMergePoint, clusterAssociation.GetUpstreamMergePoint()));
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+inline const pandora::CartesianVector ClusterAssociation::GetUpstreamMergePoint() const
+{
+    return m_upstreamMergePoint;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+inline const pandora::CartesianVector ClusterAssociation::GetUpstreamMergeDirection() const
+{
+    return m_upstreamMergeDirection;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+inline const pandora::CartesianVector ClusterAssociation::GetDownstreamMergePoint() const
+{
+    return m_downstreamMergePoint;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+inline const pandora::CartesianVector ClusterAssociation::GetDownstreamMergeDirection() const
+{
+    return m_downstreamMergeDirection;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+inline const pandora::CartesianVector ClusterAssociation::GetConnectingLineDirection() const
+{
+    return m_connectingLineDirection;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+inline void ClusterAssociation::SetUpstreamMergePoint(const pandora::CartesianVector &upstreamMergePoint)
+{
+    m_upstreamMergePoint = upstreamMergePoint;
+    
+    const pandora::CartesianVector connectingLineDirection(m_downstreamMergePoint.GetX() - m_upstreamMergePoint.GetX(), 0.f, m_downstreamMergePoint.GetZ() - m_upstreamMergePoint.GetZ());
+    m_connectingLineDirection = connectingLineDirection.GetUnitVector();
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+inline void ClusterAssociation::SetDownstreamMergePoint(const pandora::CartesianVector &downstreamMergePoint)
+{
+    m_downstreamMergePoint = downstreamMergePoint;
+    
+    const pandora::CartesianVector connectingLineDirection(m_downstreamMergePoint.GetX() - m_upstreamMergePoint.GetX(), 0.f, m_downstreamMergePoint.GetZ() - m_upstreamMergePoint.GetZ());
+    m_connectingLineDirection = connectingLineDirection.GetUnitVector();
+}
```

Comment:
I think i'll go with option 1, they are currently used in the TrackExtensionRefinementAlgorithm::IsExtrapolatedEndpointNearBoundary method. The reason for this is to provide exact endpoints of the extrapolated hits so that the IsTrackContinuous function will not fail. Although, currently I am think of removing this function altogether and if this happens i'll implement option 2...